lib/puppet/module/task.rb

```diff
@@ -33,54 +111,72 @@ def self.is_tasks_executable_filename?(name)
     end
 
     def self.tasks_in_module(pup_module)
-      Dir.glob(File.join(pup_module.tasks_directory, '*'))
+      task_files = Dir.glob(File.join(pup_module.tasks_directory, '*'))
         .keep_if { |f| is_tasks_filename?(f) }
-        .group_by { |f| task_name_from_path(f) }
-        .map { |task, files| new_with_files(pup_module, task, files) }
-    end
 
-    attr_reader :name, :module, :metadata_file, :files
+      module_executables = task_files.reject(&method(:is_tasks_metadata_filename?)).map.to_a
 
-    def initialize(pup_module, task_name, files, metadata_file = nil)
-      if !Puppet::Module::Task.is_task_name?(task_name)
-        raise InvalidName, _("Task names must start with a lowercase letter and be composed of only lowercase letters, numbers, and underscores")
+      tasks = task_files.group_by { |f| task_name_from_path(f) }
+
+      tasks.map do |task, executables|
+        new_with_files(pup_module, task, executables, module_executables)
       end
+    end
 
-      all_files = metadata_file.nil? ? files : files + [metadata_file]
-      all_files.each do |f|
-        if !f.start_with?(pup_module.tasks_directory)
-          msg = _("The file '%{path}' is not located in the %{module_name} module's tasks directory") %
-                       {path: f.to_s, module_name: pup_module.name}
+    attr_reader :name, :module, :metadata_file, :metadata
 
-          # we can include some extra context for the log message:
-          Puppet.err(msg + " (#{pup_module.tasks_directory})")
-          raise InvalidFile, msg
-        end
+    # file paths must be relative to the modules task directory
+    def initialize(pup_module, task_name,  module_executables, metadata_file = nil)
+      if !Puppet::Module::Task.is_task_name?(task_name)
+        raise InvalidName, _("Task names must start with a lowercase letter and be composed of only lowercase letters, numbers, and underscores")
       end
 
       name = task_name == "init" ? pup_module.name : "#{pup_module.name}::#{task_name}"
 
       @module = pup_module
       @name = name
-      @metadata_file = metadata_file if metadata_file
-      @files = files
+      @metadata_file = metadata_file
+      @module_executables = module_executables || []
+    end
+
+    def read_metadata(file)
+      File.open(file) { |fh| JSON.load(fh) } if file
```

Comment:
Using `JSON.load` is potentially unsafe because it can load arbitrary objects. For example, [the following](https://gist.github.com/joshcooper/5d1a619f6165d8dcbfb8e17a6f71440c), produces:

```
$ bx ruby json.rb
Foo
```

Is the `file` input trusted? If not, we should use `JSON.parse`, or pass options to disable `create_additions`.

Sorry for the late ping, I just noticed this because my PR to reenable YAML/JSON.load rubocop checks is now failling. https://github.com/puppetlabs/puppet/pull/6972

Reply:
I think `JSON.parse` would be sufficient here.

Reply:
Or maybe we should use `Puppet::Util::Json.load(Puppet::FileSystem.read(file, :encoding => 'utf-8'))`?

Reply:
makes sense, I added a commit to my pr